presto-clp/src/main/java/com/facebook/presto/plugin/clp/ClpFilterToKqlConverter.java

```diff
@@ -0,0 +1,738 @@
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.function.FunctionHandle;
+import com.facebook.presto.spi.function.FunctionMetadata;
+import com.facebook.presto.spi.function.FunctionMetadataManager;
+import com.facebook.presto.spi.function.StandardFunctionResolution;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.RowExpressionVisitor;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.slice.Slice;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.common.function.OperatorType.EQUAL;
+import static com.facebook.presto.common.function.OperatorType.GREATER_THAN;
+import static com.facebook.presto.common.function.OperatorType.GREATER_THAN_OR_EQUAL;
+import static com.facebook.presto.common.function.OperatorType.IS_DISTINCT_FROM;
+import static com.facebook.presto.common.function.OperatorType.LESS_THAN;
+import static com.facebook.presto.common.function.OperatorType.LESS_THAN_OR_EQUAL;
+import static com.facebook.presto.common.function.OperatorType.NEGATION;
+import static com.facebook.presto.common.function.OperatorType.NOT_EQUAL;
+import static com.facebook.presto.common.function.OperatorType.flip;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.plugin.clp.ClpErrorCode.CLP_PUSHDOWN_UNSUPPORTED_EXPRESSION;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.AND;
+import static java.lang.Integer.parseInt;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * ClpFilterToKqlConverter translates Presto RowExpressions into KQL (Kibana Query Language) filters
+ * used as CLP queries. This is used primarily for pushing down supported filters to the CLP engine.
+ * This class implements the RowExpressionVisitor interface and recursively walks Presto filter expressions,
+ * attempting to convert supported expressions (e.g., comparisons, logical AND/OR, LIKE, IN, IS NULL,
+ * and SUBSTR-based expressions) into corresponding KQL filter strings. Any part of the expression that
+ * cannot be translated is preserved as a "remaining expression" for potential fallback processing.
+ * Supported translations include:
+ * - Variable-to-literal comparisons (e.g., =, !=, <, >, <=, >=)
+ * - String pattern matches using LIKE
+ * - Membership checks using IN
+ * - NULL checks via IS NULL
+ * - Substring comparisons (e.g., SUBSTR(x, start, len) = "val") mapped to wildcard KQL queries
+ * - Dereferencing fields from row-typed variables
+ * - Logical operators AND, OR, and NOT
+ */
```

Comment:
```suggestion
/**
 * A translator to translate Presto RowExpressions into KQL (Kibana Query Language) filters used as
 * CLP queries. This is used primarily for pushing down supported filters to the CLP engine. This
 * class implements the RowExpressionVisitor interface and recursively walks Presto filter
 * expressions, attempting to convert supported expressions into corresponding KQL filter strings.
 * Any part of the expression that cannot be translated is preserved as a "remaining expression" for
 * potential fallback processing.
 * <p></p>
 * Supported translations include:
 * <ul>
 *     <li>Comparisons between variables and literals (e.g., =, !=, <, >, <=, >=)</li>
 *     <li>String pattern matches using LIKE (some patterns are unsupported)</li>
 *     <li>Membership checks using IN</li>
 *     <li>NULL checks via IS NULL</li>
 *     <li>Substring comparisons (e.g., SUBSTR(x, start, len) = "val")</li>
 *     <li>Dereferencing fields from row-typed variables</li>
 *     <li>Logical operators AND, OR, and NOT</li>
 * </ul>
 */
```

Can we also mention we can only do comparisons between variable & constant but not variable & variable?

Reply:
Sounds good.